### PR TITLE
memory viewer 

### PR DESCRIPTION
**Pull request by idle-PB**
Added Decimal Octal Hex views to the table data view, via a button that cycles the mode and a checkbox data to optionally format the output for a data section on copy or save. 
Added ~700 ASM mnemonics, they will also have to be added to the compiler. 
added a Linux-x64.sh build with instructions on how to compile but it's dirty as it requires copying the gtk2 subsystem libs over the gtk3 libs otherwise it will link both and won't run.

### CODE REVIEW

File: PureBasicDebugger/MemoryViewer.pb [hunk 1]
```diff
@@ -3,23 +3,128 @@
...
+      ProcedureReturn "DEC" 
+    Case #MEMORY_VIEW_TABLE_DATA_HEX
+      ProcedureReturn "HEX"
+    Case #MEMORY_VIEW_TABLE_DATA_OCT 
+      ProcedureReturn "OCT"
+  EndSelect     
+EndProcedure 
+
+Procedure.s OCT(number.q)
+  Protected oct.s=Space(8*SizeOf(Character))
```
Comment (fantaisie-software) [inline PureBasicDebugger/MemoryViewer.pb]: Are you sure you need to multiply with SizeOf(Character) ? Space() already takes NbChar as parameter

File: PureBasicDebugger/MemoryViewer.pb [hunk 2]
```diff
@@ -3,23 +3,128 @@
...
+  For a = 7 To 0 Step -1
+    PokeS(@oct+a*SizeOf(Character),Str(number & 7),SizeOf(Character),#PB_String_NoZero)
+    number >> 3
+  Next 
+  oct = Trim(oct,"0") 
+  If oct = "        "
```
Comment (fantaisie-software) [inline PureBasicDebugger/MemoryViewer.pb]: The whole IDE use uppercase as first variable name, would be great to continue like that :)

File: PureBasicDebugger/MemoryViewer.pb [hunk 3]
```diff
@@ -43,7 +148,7 @@ Procedure MemoryViewer_Table(*Debugger.DebuggerData, VariableSize, PeekVal.Memor
     
     For i = 0 To columns-1
       AddGadgetColumn(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_List], i+1, Hex(i * VariableSize), width)
-    Next i
+    Next i
```
Comment (fantaisie-software) [inline PureBasicDebugger/MemoryViewer.pb]: why the extra space at line end ?

File: PureBasicDebugger/MemoryViewer.pb [hunk 4]
```diff
@@ -269,8 +374,9 @@ Procedure MemoryViewer_Unicode(*Debugger.DebuggerData)
 EndProcedure
 
 Procedure MemoryViewer_Update(*Debugger.DebuggerData, Action, File) ; 0=display only, 1=display+copy to clipboard, 2=display+save to file
-  
+  Protected datatype.s, OptExportDS=1 
   ViewType = GetGadgetState(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_ViewType])
```
Comment (fantaisie-software) [inline PureBasicDebugger/MemoryViewer.pb]: datatype -> DataType

File: PureBasicDebugger/MemoryViewer.pb [hunk 5]
```diff
@@ -479,6 +616,7 @@ Procedure MemoryViewerWindowEvents(*Debugger.DebuggerData, EventID)
     TextWidth    = Max(65, GetRequiredWidth(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_Text]))
     
     ButtonHeight = Max(ButtonHeight, GetRequiredHeight(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_ViewType]))
+    dataWidth    = MAX(100, GetRequiredWidth(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_ChkformatDataSection]))
```
Comment (fantaisie-software) [inline PureBasicDebugger/MemoryViewer.pb]: DataWidth

File: PureBasicDebugger/MemoryViewer.pb [hunk 6]
```diff
@@ -633,6 +777,8 @@ Procedure UpdateMemoryViewerWindow(*Debugger.DebuggerData)
   SetWindowTitle(*Debugger\Windows[#DEBUGGER_WINDOW_Memory], Language("Debugger","MemoryWindowTitle") + " - " + GetFilePart(*Debugger\FileName$))
   SetGadgetText(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_Text], Language("Debugger","Range")+":")
   SetGadgetText(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_Display], Language("Debugger","Display"))
+  SetGadgetText(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_Display_DataView],MemoryViewer_GetDataView(MemoryViewTableData))
+  SetGadgetText(*Debugger\Gadgets[#DEBUGGER_GADGET_Memory_ChkformatDataSection],"Data")
```
Comment (fantaisie-software) [inline PureBasicDebugger/MemoryViewer.pb]: DEBUGGER_GADGET_Memory_ChkformatDataSection -> DEBUGGER_GADGET_Memory_CheckFormatDataSection

File: PureBasicDebugger/MemoryViewer.pb (hunk 1 above)
Comment (idle-PB) [reply PureBasicDebugger/MemoryViewer.pb]: #DEBUGGER_GADGET_Memory_ChkformatDataSection needs to be added into DebuggerCommon.pb it got wiped messing around with git which also broke my build. 

no probably don't need the * sizeOf(Character) either in that case, I'll look into it.

File: PureBasicDebugger/MemoryViewer.pb (hunk 2 above)
Comment (idle-PB) [reply PureBasicDebugger/MemoryViewer.pb]: Ok I will try

**Comment by kenmo-pb** [comment]
I have not tested this yet, but there are some odd things in the diff / Files Changed.

There are some lines that are just whitespace changes (spaces added at end of line?) and some comment typos are reintroduced (is there a mix-up of old and new file versions again?)

**Comment by idle-PB** [comment]
git is driving me crazy, the memory viewer wasn't the right file, I just re uploaded it along with debuggercommon so hopefully that will sort it out but I fear, it will only stuff it up more.

**Comment by fantaisie-software** [comment]
Don't worry to much idle-PB, we all been experienced hard debut with git haha :). Thanks for resubmitting the PR, I will take a look

**Comment by idle-PB** [comment]
as an example  428 in Dec : 654 in Oct : $1AC in Hex 
fpem.u = 428
ShowMemoryViewer(@fpem,2)

and yes maybe a combo box would be better

**Comment by idle-PB** [comment]
changed to combo box, see previous comment about how it should look.

**Comment by fantaisie-software** [comment]
Did you pushed it ? I can't see the combo. I see what you mean now, but it only works in 'table' mode right ? If we stay in 'Hex View', it has no effect. If so, you should gray it in 'Hex View' and enable it when the view is switched to a table. You could even move the combo next to the view selection as it's directly related.

**Comment by idle-PB** [comment]
I committed it but it didn't appear where it should, then I tried to append it but that didn't work out either and then I just added it to a branch memview. I'm still trying to figure out git / github. 

And just now I cocked it up again it appears on my master not the branch!  
I moved it next to the viewtype combo and set it to enabled and disabled based on the viewtype

**Comment by idle-PB** [comment]
looks like I'll have to close this pull request and then do another from the memview branch I created.

**Comment by tajmone** [comment]
@idle-PB, remember that I've kept a backup of the previous PR (plus my UTF-8 fix):

https://github.com/tajmone/purebasic/tree/idle-PB-fix-memory-viewer

... in case you need it (to recover the files).

**Comment by idle-PB** [comment]
ok thanks but can I close this pull request and re do it from the memview branch?

**Comment by idle-PB** [comment]
created a new pull from memview branch since I can't see how  to append the changes to this.

**Comment by tajmone** [comment]
> created a new pull from memview branch

You may want to close this PR then?

**Pull request closed without merging**